### DISCUSSION TITLE
Use `kind` rather than `type` as discriminator in MF2 JSON data model

### DISCUSSION THREAD

**Discussion by lucacasonato** [unicode-org/message-format-wg#628] [Feedback]
Many languages use `type` as a keyword, which makes naming fields of (deserialised) structs `type` difficult. An example of this is Rust, where naming struct properties `type` is not possible without doing a fun dance of prefixing the field with `r#`.

Right now MF2 uses `"type"` as the discriminator in the JSON data model.

Can I suggest you use `"kind"` as the name for this discriminator to avoid this inconvenience? `kind` is generally not used as a keyword in programming languages.

**Comment by eemeli** [comment]
One issue with `kind` is that we're already using it, for [markup](https://github.com/unicode-org/message-format-wg/tree/main/spec/data-model#markup).

Regarding `type`, my expectation would be that in languages like Rust the JSON representation of each node would be marshaled into some struct, which would not need a `"type"` field. Is that not a reasonable presumption?

`type` is also intentionally chosen to match the common practices in JavaScript and other ASTs, though that's not universal.

**Reply by lucacasonato** [reply]
Yes, it is possible that in Rust you would marshal into an enum for example. In this case the point becomes moot.

However, it's not so simple in other languages. In Go for example, `type Foo struct { type string }` is not valid.

I think it may also be confusing to have both a `"kind"` and `"type"` on the same object. Maybe a different name can be found for the markup `"kind"`? Maybe `"position"` (`"position": "open"`, `"position": "close"` etc)?

I think `"kind"` is also not necessarily less common than `"type"` in other ASTs. A difference here from for example the JS ASTs expressed in JS, is that this AST format is designed to be used in many different languages, not just JS. As such, I think it's quite important that the format is easially portable to other languages. Part of making it feel "native" to more than just JS, is that you don't have to use weird incantations like `r#type`, or `type_` in non JS languages.
